### PR TITLE
TOK-516: document the contracts

### PR DESCRIPTION
**Pull request by antomor**
## What

- document the contracts included in the repo
- include a specification file to describe how the rewards are calculated

### CODE REVIEW

File: specs/CONTRACTS.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+One `Gauge` is deployed for each Builder who is approved by the community. The Gauge is in charge of receiving the rewards to be distributed between the Builder and the Backers. It's also responsible for keeping track of the Backers' allocation. The Builder is always identified with a unique address and there is always one gauge associated to one Builder.
+
+## GaugeBeacon
+
+To create gauges for each Builder, we made use of the [Beacon proxy pattern](https://docs.openzeppelin.com/contracts/3.x/api/proxy#beacon). GaugeBeacon is the contract in charge of keeping track of the implementation.
+
+## GaugeFactory
+
+GaugeFactory is the contract which creates Gauge when a new Builder is approved by the community. It makes use of the GaugeBeacon to create new [BeaconProxy](https://docs.openzeppelin.com/contracts/3.x/api/proxy#BeaconProxy) contracts.
```
Comment (jmendiola222) [inline specs/CONTRACTS.md:15]:
```suggestion
GaugeFactory is the contract which creates a Gauge when a new Builder is approved by the community. It makes use of GaugeBeacon to create new [BeaconProxy](https://docs.openzeppelin.com/contracts/3.x/api/proxy#BeaconProxy) contracts.
```

File: specs/CONTRACTS.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+## BackersMangers
+
+This contract is in charge of:
+- allocating Backers votes to Builder gauges. Backers can allocate to one Builder or multiple Builders at the same time.
+- managing the distributution of the rewards across the Builder gauges
+- allow the Backers to claim all of the rewards in a single transaction.
```
Comment (jmendiola222) [inline specs/CONTRACTS.md:22]:
```suggestion
- allowing Backers to claim all of their rewards in a single transaction.
```

File: specs/CONTRACTS.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+## GaugeBeacon
+
+To create gauges for each Builder, we made use of the [Beacon proxy pattern](https://docs.openzeppelin.com/contracts/3.x/api/proxy#beacon). GaugeBeacon is the contract in charge of keeping track of the implementation.
+
+## GaugeFactory
+
+GaugeFactory is the contract which creates Gauge when a new Builder is approved by the community. It makes use of the GaugeBeacon to create new [BeaconProxy](https://docs.openzeppelin.com/contracts/3.x/api/proxy#BeaconProxy) contracts.
+
+## BackersMangers
```
Comment (franciscotobar) [inline specs/CONTRACTS.md:17]:
```suggestion
## BackersManagers
```

File: specs/SPECIFICATION.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+The Collective Rewards protocol aims to create a decentralized system to allow Builders and Backers to get rewards. The main idea is to allow Backers who have StRIF (by staking RIF), to allocate their votes to activated Builders. In order for a Builder to be activated they need to:
+- get the KYC approval by the KYC Approver
+- get the community approval through the RootstockCollective DAO.
+When a Builder is KYC approved, they need to specify the amount of rewards that they plan to distribute to their Backers as a percentage of the rewards that they'll get.
+Once the Builder is activated, they can be voted by Backers. The number of votes that a Builder gets will determine the rewards the Builder will be entitled of. If a Backer allocated votes for a Builder, they'll receive rewards that are proportional to the votes allocated and the amount of time for which the votes stay allocated to that Builder.
+
+Example:
+Alice is a Builder.
+Bob and Chad are Backers.
+Alice is KYC and community approved and her backer rewards percentage is 50%.
```
Comment (jmendiola222) [inline specs/SPECIFICATION.md:30]: I don't quite get this part:
> Alice is KYC and community approved and her backer rewards percentage is 50%.
> The Backers who vote for Alice will get 25% of the rewards distributed to Alice.
If the rewardPerc is 50%, why Backers get 25%?

File: specs/SPECIFICATION.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+### Scenario 1: single Backer
+
+A Gauge has 1000 tokens to distribute to Backers over a cycle of 100 seconds.
+The leftover is 0 because the reward rate is initially 0.
+The reward missing is initially 0.
+The reward per token stored is currently 0, because no time has passed.
+The reward rate at time 0 is 1000/100 = 10. 10 tokens will be distributed for each second.
+Last time update = 0.
+
+Alice allocates 100 votes at time 10.
```
Comment (jmendiola222) [inline specs/SPECIFICATION.md:109]: In previos example, we used Alice as Builder and Bob & Chand as backers, I'll try to keep the same personas.

File: specs/SPECIFICATION.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+Last time update = 100.
+Alice rewards = old value + (Alice allocation * (reward per token stored - Alice reward per token paid)) = 100 * (7.3 - 0) = 730.
+Alice per token paid is set to reward per token stored that is equal to 7.3.
+Token is transferred to Bob address
+Alice reward is set to 0.
+
+
+## Builder incentivization
+
+In order for the Builder to distribute rewards, they need to earn rewards. Builders could also incentivize votes by allocating Rewards on Gauge contracts by means of `Gauge#incentivizeWithRewardToken()` and `Gauge#incentivizeWithCoinbase()`. Those rewards will be available only for the Backers who vote for the Builder and they'll be undistinguishable between the rewards gained by votes.
```
Comment (jmendiola222) [inline specs/SPECIFICATION.md:171]: > In order for the Builder to distribute rewards, they need to earn rewards.
Isn't this confusing as we are explainig an alternative mechanism to distribute rewards that doesn't require earning?

File: specs/CONTRACTS.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+- approve Builders KYC
+- revoke Builders KYC
+- community approve Builders
+- remove community approval for the Builders
+- pause/unpause Builders
+- store Backers reward percentage
+- keep track of the Builder<->Gauge association
+- allow the Builder to change the Backers reward percentage
+- allow the Builder to replace the Builder reward receiver address
+- get the list of gauges
```
Comment (franciscotobar) [inline specs/CONTRACTS.md:38]: We are missing to revoke themselves isn't it?

File: specs/CONTRACTS.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+## CycleTimeKeeper
+
+It's in charge of managing the cycle and the distribution and to retrieve the related information:
```
Comment (franciscotobar) [inline specs/CONTRACTS.md:42]:
```suggestion
It's in charge of managing the cycle and the distribution time and to retrieve the related information:
```

File: specs/CONTRACTS.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+
+## GovernanceManager
+
+Its responsibility is to manage all the roles involved in the protocol and to execute complex changes by means of an authorized changer. It exposes few functions to validate that an address is one of the managed roles and to update the addresses assigned to those roles.
+- governor: the contract representing the DAO governor TODO: add link here
+- authorized changer: a smart contract temporarily authorized by the Governor
+- foundationTreasury: an address in charge of managing the RewardDistributor
+- kycApprover: an address whose responsibility is to manage the KYC
+- upgrader: ad address that is authorized (together with the governor and an authorized changer) to upgrade the contracts
```
Comment (franciscotobar) [inline specs/CONTRACTS.md:62]:
```suggestion
- upgrader: an address that is authorized (together with the governor and an authorized changer) to upgrade the contracts
```

File: specs/SPECIFICATION.md (hunk 4 above)
Comment (antomor) [reply specs/SPECIFICATION.md:30]: Not sure why, probably my initial example was different 👍🏽

File: specs/SPECIFICATION.md (hunk 5 above)
Comment (antomor) [reply specs/SPECIFICATION.md:109]: Good point, I've changed the first example for simplicity

**Review by jmendiola222 — APPROVED** [review]
LGTM

**Pull request merged**